localization/localization_util/src/smart_pose_buffer.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023- Autoware Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "localization_util/smart_pose_buffer.hpp"
+
+SmartPoseBuffer::SmartPoseBuffer(
+  const rclcpp::Logger & logger, const double & pose_timeout_sec,
+  const double & pose_distance_tolerance_meters)
+: logger_(logger),
+  pose_timeout_sec_(pose_timeout_sec),
+  pose_distance_tolerance_meters_(pose_distance_tolerance_meters)
+{
+}
+
+std::optional<SmartPoseBuffer::InterpolateResult> SmartPoseBuffer::interpolate(
+  const rclcpp::Time & target_ros_time)
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (pose_buffer_.size() < 2) {
+    RCLCPP_INFO(logger_, "pose_buffer_.size() < 2");
+    return std::nullopt;
+  }
+
+  const rclcpp::Time time_first = pose_buffer_.front()->header.stamp;
+  const rclcpp::Time time_last = pose_buffer_.back()->header.stamp;
+
+  if (target_ros_time < time_first || time_last < target_ros_time) {
+    return std::nullopt;
+  }
+
+  InterpolateResult result;
+
+  // get the nearest poses
+  result.old_pose = *pose_buffer_.front();
+  for (const PoseWithCovarianceStamped::ConstSharedPtr & pose_cov_msg_ptr : pose_buffer_) {
+    result.new_pose = *pose_cov_msg_ptr;
+    const rclcpp::Time pose_time_stamp = result.new_pose.header.stamp;
+    if (pose_time_stamp > target_ros_time) {
+      break;
+    }
+    result.old_pose = *pose_cov_msg_ptr;
+  }
+
+  // check the time stamp
+  bool is_old_pose_valid = validate_time_stamp_difference(
+    result.old_pose.header.stamp, target_ros_time, pose_timeout_sec_);
+  bool is_new_pose_valid = validate_time_stamp_difference(
+    result.new_pose.header.stamp, target_ros_time, pose_timeout_sec_);
+
+  // check the position jumping (ex. immediately after the initial pose estimation)
+  bool is_pose_diff_valid = validate_position_difference(
+    result.old_pose.pose.pose.position, result.new_pose.pose.pose.position,
+    pose_distance_tolerance_meters_);
```

Comment:
You can make these variables const variables.

Reply:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5713/commits/cbfb8bfb417da20a4a9bacfa01f7e321683c260b